### PR TITLE
Run3-hcx300 Update code for new shower library (Lev & Salavat)

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Update code for new shower library (Lev & Salavat), Uses a modifier run3_HFSL to activate the new library. This can be activated only when the corresponding file is saved in the data repository

#### PR validation:

Use the runTheMatrix test workflow and a new script in SimG4CMS/Calo/test/python/runHF6_cfg.py

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Nothing special

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by cvuosalo** [comment]
+1

File: SimG4CMS/Calo/src/HFCherenkov.cc [hunk 1]
```diff
@@ -167,32 +132,45 @@ int HFCherenkov::computeNPE(const G4Step* aStep,
...
           wlatten.push_back(lambda);
           rand = G4UniformRand();
           double effHEM = computeHEMEff(lambda);
+          // compute number of bounces in air guide
+          rand = G4UniformRand();
+          double phi = 0.5 * M_PI * rand;
+          double rad_bundle = 19.;  // bundle radius
+          double rad_lg = 25.;      //  light guide radius
+          rand = G4UniformRand();
+          double rad = rad_bundle * sqrt(rand);
```
Comment (civanch) [inline SimG4CMS/Calo/src/HFCherenkov.cc:161]: Better to use std::sqrt()

File: SimG4CMS/Calo/src/HFCherenkov.cc [hunk 2]
```diff
@@ -17,63 +17,28 @@
 
 HFCherenkov::HFCherenkov(edm::ParameterSet const& m_HF) {
   ref_index = m_HF.getParameter<double>("RefIndex");
-  lambda1 = ((m_HF.getParameter<double>("Lambda1")) / pow(double(10), 7)) * cm;
-  lambda2 = ((m_HF.getParameter<double>("Lambda2")) / pow(double(10), 7)) * cm;
-  aperture = cos(asin(m_HF.getParameter<double>("Aperture")));
-  apertureTrap = cos(asin(m_HF.getParameter<double>("ApertureTrapped")));
-  aperturetrapped = m_HF.getParameter<double>("CosApertureTrapped");
+  lambda1 = ((m_HF.getParameter<double>("Lambda1")) / pow(double(10), 7)) * CLHEP::cm;
```
Comment (civanch) [inline SimG4CMS/Calo/src/HFCherenkov.cc:20]: @bsunanda , for me it is strange to use pow(double(10), 7)), instead I would put 1.e+7

File: SimG4CMS/Calo/src/HFFibre.cc [hunk 3]
```diff
@@ -1,151 +1,152 @@
...
+
+HFFibre::HFFibre(const std::string& name,
+                 const HcalDDDSimConstants* hcons,
+                 const HcalSimulationParameters* hps,
+                 edm::ParameterSet const& p)
+    : hcalConstant_(hcons), hcalsimpar_(hps) {
+  edm::ParameterSet m_HF =
+      (p.getParameter<edm::ParameterSet>("HFShower")).getParameter<edm::ParameterSet>("HFShowerBlock");
+  cFibre = c_light * (m_HF.getParameter<double>("CFibre"));
+#ifdef EDM_ML_DEBUG
```
Comment (civanch) [inline SimG4CMS/Calo/src/HFFibre.cc:24]: @bsunanda , #ifdef is useful in run time code but not in class constructor.

**Comment by civanch** [comment]
@bsunanda , I make few comments, which are really minor but there is one , which I do not yet understand: switch to new SL happens in g4SimHits_cfi, why you put in also in scripts of this PR inside HFParameters?

**Comment by abdoulline** [comment]
Good news:
with/without  this PR a standard 2021 (run3) test [1] without any customization yields idential results.
Bad news:
with this PR and added customization [2] to acivate it with new SL, the _new_ results are not good at all  [3].

---------------------------------------------------------
[1]
process = cms.Process('TEST',eras.Run3)
process.load("Configuration.StandardSequences.GeometryDB_cff")  #  standard Geometry

[2]

process.HFLibraryFileBlock.FileName = 'HFShowerLibrary_run3_v5.root' #  in /src  locally
process.HFLibraryFileBlock.FileVersion = 1
process.HFShowerBlock.IgnoreTimeShift   = True

+ timePhase adjustment for new PCaloHits timing 37-38 ns (for electros)  instead of ~52-54 in the default vestion 

[3]
 https://cms-cpt-software.web.cern.ch/cms-cpt-software/General/Validation/SVSuite/HCAL/calo_scan_single_pi/12_X_newSL_PR34432/12_0_X_run3_PR34432_newSL_tP-1_vs_12_0_X_run3_SinglePi/

->  PCaloHits sum in R=0.3 vs ieta  (without any time cuts): 
https://cmssdt.cern.ch/lxr/source/Validation/HcalHits/src/HcalSimHitsValidation.cc#0474    :

https://cms-cpt-software.web.cern.ch/cms-cpt-software/General/Validation/SVSuite/HCAL/calo_scan_single_pi/12_X_newSL_PR34432/12_0_X_run3_PR34432_newSL_tP-1_vs_12_0_X_run3_SinglePi/SimHitsTask_emean_vs_ieta_EH.gif

**Comment by bsunanda** [comment]
@civanch Shower library is used for full and fast simulation. That is why the blocks are defined outside the domain of Full/FastSim

**Comment by civanch** [comment]
@bsunanda , sorry, I forget this discussion and this move. If g4SimHits peak up central parameters why have some parameters and modifier inside g4SimHits and not include the file?

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by abdoulline** [comment]
@bsunanda  Thanks for the fix!  
Now low-level PCaloHits tests show reasonable results for new HF ShowerLibrary.

**Comment by abdoulline** [comment]
Good news afther the fix:
(1) PCaloHits look reasonable now both for electrons and pions
( 2) I've quickly/approximately reset relevant HF Digitization parameters (timePhase(s) and samplingFactor(s)) and now standard 50 GeV pion-gun test **with new SL** yields very goos agreement with the default SL (actual MC):
https://cms-cpt-software.web.cern.ch/cms-cpt-software/General/Validation/SVSuite/HCAL/calo_scan_single_pi/12_X_newSL_PR34432/12_0_X_run3_PR34432_newSL_tP-1_SF0.36_vs_12_0_X_run3_SinglePi/

**Comment by cvuosalo** [comment]
+1

File: SimG4CMS/Calo/src/HFCherenkov.cc (hunk 1 above)
Comment (bsunanda) [reply SimG4CMS/Calo/src/HFCherenkov.cc]: Taken care of

File: SimG4CMS/Calo/src/HFCherenkov.cc (hunk 2 above)
Comment (bsunanda) [reply SimG4CMS/Calo/src/HFCherenkov.cc]: Taken care of

**Comment by bsunanda** [comment]
@civanch Please approve

**Comment by civanch** [comment]
+1

**Comment by qliphy** [comment]
+1

**Comment by davidlange6** [comment]
belatedly - has someone assessed the performance impact of this new library in the context of the parameters with which  the ROOT file was created?

**Comment by davidlange6** [comment]
Its not really in the shadow is it? Maybe someone has a recent profile

**Comment by civanch** [comment]
@davidlange6 , recently we did profile with Run2 shower library in 12_0_0_pre1 and it was not contributing to CPU in a significant way. We should repeat the exercise when this new SL will be integrated.

**Comment by davidlange6** [comment]
Anyway, in what way is the new library compressed?

**Comment by abdoulline** [comment]
@davidlange6   David, the situation is as follows: 

(1) you're right, there seems to be sizable impact of using new SL (HF SHowerLibrary) on SIM performnce:
  ~25% slow down of GENSIM step for 11634.0 wf [*];

(2) files used for writing components (electrons, pions for given energies) of new SL and merging them in final SL,
alas, were not committed to CMSSW (unlike all other codes related to new SL) and wer used in private by Lev Kheyn, please find them in his public 
 /afs/cern.ch/user/k/kheyn/public/For_Sunanda     - 
  (a)  HFShowerLibraryAnalyzer.cc   - code reading "FullSim" (Full SImulation of Cherenkov light) root files and creating library
  (b)  writelibraryfile_cfg.py  - config
  (c)  fileList.txt -  used list of FullSim files

We'd appreciate your expert's insight - whether something can be done to improve the situation (1)... 

----------------------------------------------------------------------
[*] Quick and (very) dirty comparison of CPU performance by running _simultaneously_ (in parallel) two jobs **w/wo new SL** for wf 11634.0 (noPU TTbar 14 TeV 2021, 100 ev step1=GENSIM)  on the same lxplus PC, numbers coming out of cms.Service("Timing")

**Comment by davidlange6** [comment]
My guess would be that you want 
1 a faster compression algorithm (eg, reduce 9 to 4 or perhaps use zstandard).
2 smaller baskets as this file basically gets read as random access. 
3 its not obvious that you want this file to be an EDM file.

**Comment by abdoulline** [comment]
@davidlange6 
could you, please, convert/translate your consideration into a prescription/recipe 
for aforementioned (2)  files  in /afs/cern.ch/user/k/kheyn/public/For_Sunanda/  ?

**Comment by davidlange6** [comment]
I’ll let the experts have a look . At least the first two can be found by reading documentation found at
edmPluginHelp -p PoolOutputModule.

**Comment by abdoulline** [comment]
@davidlange6 
Hi David,  sorry to bug you once again,  unfortunately we were misled by the files put by Lev in a rush to the aformentioned public directory, who has produced/assembled new SL [1].
Actually he didn't use PoolOutputModule, but TFileService... 

I've published (somewhat modified version of) both code and job config for assembling SL from 32 input components  in  https://github.com/cms-sw/cmssw/pull/34561   and produced yet another version, after many attempts to improve the performance of new SL(s) in TTbar 2021 MC SIM step. 
NB: new SL contains _twice_ more number of showers per each energy point, so its size is inevitably bigger than for actual default  ().
Finally the best updated new SL has been found/produced  with parameters buffersize=1 and splitlevel=2 for Tree->Branches. 
This modified/updated version of new SL is ~5% slower for TTbar 2021 MC SIM wrt actual default SL, but not ~25% anymore as for the intial version produced by Lev with ROOT defaults for Tree->Branch  buffersize=32000 and splitlevel=99.
Do you see any further optimization possible? 

Unfortunately we have little idea what did you do 5 years ago (2016), when you've somehow reformatted initial version of HFShowerLibrary_npmt_noatt_eta4_16en_v4.root (produced by Lev back in 2015) so that it became smaller (~500 MB -> 330 MB) and probably faster (no numbers available though...) ? 

----------------------------------------------------------------------------------
[1] 
HFShowerLibrary_run3_v5.root,  put by Shahzad into  https://cmssdt.cern.ch/SDT/data/CMSSW/SimG4CMS/Calo/data/

[2]

**Comment by abdoulline** [comment]
@davidlange6 
may be the folowing explicit settings may further help :    
  fs->file().SetCompressionAlgorithm(4);
  fs->file().SetCompressionLevel(4);
when added here
https://cmssdt.cern.ch/lxr/source/SimG4CMS/ShowerLibraryProducer/plugins/HcalForwardLibWriter.cc#0020

**Comment by abdoulline** [comment]
Update: 

Lev has reported (using single-particle gun, shot at HF) that the latest/biggest SL is the best (as evaluated on OscarMTProducer TimeService fragments collected over 100 ev) among three new tried SL's, ~60-70 times faster (per single particle) than aforementioned cmssdt-residing [1],  but still not at the level of the actual default one...

**Comment by davidlange6** [comment]
These sound like the right sort of changes. I understand the new library is 3x bigger than the run 2 library - naturally that will be slower to access (as the code jumps around in the library). Is it already single precision? Beyond that, you might compare the split level of the old file (TTree::splitlevel())

**Comment by abdoulline** [comment]
Finally Lev has re-done his OscarMTProducer CPU time eveluations with single 100 GeV pions and it seems that the "lates-geratest" version of new SL is the best of all compared (left to right) 

CPU_time_OSCAR_pi100_SL_versions

(1)  green : new SL with added (wrt (3) below) LZ4 compression algo and compression level=4 
     now in /afs/cern.ch/user/a/abdullin/public/updated_HF_SL_July22_2021/HFShowerLibrary_run3_v5.root
(2) black: actual default HFShowerLibrary_npmt_noatt_eta4_16en_v4.root
(3) blue: my intermediate attempt to re-assemble new SL only using   buffersize=1 and splitlevel=2 for Tree->Branches
     https://github.com/cms-sw/cmssw/pull/34432#issuecomment-884676611
(4) red: HFShowerLibrary_run3_v5(_210713).root   - the initial version of new SL committed to CMSSDT

**Pull request merged**